src/parser/expr.rs

```diff
@@ -1,122 +1,284 @@
-    // TODO: this should go to comparison
-    fn inversion(&mut self) -> ParseResult<Expr<'s, 'a>> {
-        if let Some(span) = self.accept_optional(tok![!])? {
-            let arg = self.inversion()?;
-            let span = self.spans.store_merged(span, &arg);
-            Ok(ExprKind::Not(self.alloc(arg)).with_span(span))
-        } else {
-            self.sum()
+    /// Parse a binary expression.
+    ///
+    /// # Grammar
+    ///
+    /// ```peg
+    /// bin_expr = una_expr (bin_op bin_expr)*
+    /// ```
+    fn bin_expr(&mut self, prev: Option<(BinOp, Span)>) -> ParseResult<Expr<'s, 'a>> {
+        // The expression parsed thus far.
+        let mut expr = self.una_expr()?;
+
+        // Try extending into a larger binary operation.
+        while let Some(next) = self.bin_op(prev)? {
+            // Get the right-hand side of this operation.
+            let rhs = self.bin_expr(Some(next))?;
+
+            // Construct the new binary operation.
+            let span = self.spans.merge(&expr, &rhs);
+            let [lhs, rhs] = [expr, rhs].map(|x| self.alloc(x));
+            expr = next.0.into_expr(lhs, rhs).with_span(span);
         }
+
+        Ok(expr)
     }
 
-    // sum = sum '+' term
-    //     | sum '-' term
-    //     | term
-    fn sum(&mut self) -> ParseResult<Expr<'s, 'a>> {
-        let mut left = self.term()?;
-        loop {
-            if self.accept_optional(tok![+])?.is_some() {
-                let right = self.term()?;
-                let span = self.spans.merge(&left, &right);
-                left = ExprKind::Add(self.alloc(left), self.alloc(right)).with_span(span);
-            } else if self.accept_optional(tok![-])?.is_some() {
-                let right = self.term()?;
-                let span = self.spans.merge(&left, &right);
-                left = ExprKind::Sub(self.alloc(left), self.alloc(right)).with_span(span);
-            } else {
-                return Ok(left);
+    /// Parse a binary operator, with precedence checking.
+    ///
+    /// # Grammar
+    ///
+    /// ```peg
+    /// bin_op = '+' | '-' | '*' | '/'
+    ///        | '==' | '!=' | '<' | '<=' | '>' | '>='
+    ///        | '&&' | '||'
+    /// ```
+    fn bin_op(&mut self, lhs: Option<(BinOp, Span)>) -> ParseResult<Option<(BinOp, Span)>> {
+        let operators = [
+            BinOp::Add,
+            BinOp::Sub,
+            BinOp::Mul,
+            BinOp::Div,
+            BinOp::IsEQ,
+            BinOp::IsNE,
+            BinOp::IsLT,
+            BinOp::IsLE,
+            BinOp::IsGT,
+            BinOp::IsGE,
+            BinOp::Con,
+            BinOp::Dis,
+        ];
+
+        // Parse the actual operator.
+        let Some(rhs) = self
+            .peek()?
+            .and_then(|token| operators.into_iter().find(|o| o.token() == token))
+        else {
+            return Ok(None);
+        };
+
+        // If there's no left-side operator, return immediately.
+        let Some(lhs) = lhs else {
```

Comment:
Why would there ever be no left hand side?

Reply:
With a top-level expression, or within parentheses, there's no left-hand side operator deciding precedence - you just parse everything.

Reply:
Feels like that could maybe be a separate function? I.e. parse the first bit and then call this function. Not sure if that would be cleaner

Reply:
Then parsing for top-level expressions would be separate from parsing for non-top-level expressions - but the core logic (everything in `binary_expr()`) would basically be the same, so we'd just duplicate that function.  I don't expect to have any performance issues with it either, so I don't think it's worth it.

Reply:
The main issue is really that `binary_op()` would have to be duplicated too - one version with precedence checking, and the other version without.  It just looks like too many new items to be worth it.